Introduce field type INTERVAL to BOX

ImeevMA opened the pull request:
This patch-set creates a new MsgPack type MP_INTERVAL and then introduces a new field type INTERVAL.

Gerold103 commented on third_party/lua-cjson/lua_cjson.c:
```diff
@@ -439,6 +439,13 @@ static void json_append_data(lua_State *l, struct luaL_serializer *cfg,
+        {
+            char buf[DT_IVAL_TO_STRING_BUFSIZE];
+            interval_to_string(field.interval, buf, sizeof(buf));
+            size_t sz = strlen(buf);
```
Can't it be the same as datetime code above? Without `strlen()`.

ImeevMA replied:
Fixed.

Gerold103 commented on third_party/lua-yaml/lyaml.cc:
```diff
@@ -716,6 +716,11 @@ static int dump_node(struct lua_yaml_dumper *dumper)
+      case MP_INTERVAL:
+         interval_to_string(field.interval, buf, sizeof(buf));
+         str = buf;
+         len = strlen(str);
```
Ditto.

ImeevMA replied:
Fixed.

Gerold103 commented on third_party/lua-yaml/lyaml.cc:
```diff
@@ -618,7 +618,7 @@ static int dump_node(struct lua_yaml_dumper *dumper)
    yaml_scalar_style_t style = YAML_PLAIN_SCALAR_STYLE;
    int is_binary = 0;
-   char buf[MAX(FPCONV_G_FMT_BUFSIZE, DT_TO_STRING_BUFSIZE)];
+   char buf[DT_IVAL_TO_STRING_BUFSIZE];
```
I would suggest to define a new macro as max of these 3 values. Otherwise you can't guarantee the buffer will always fit all 3 types if the sizes will change in the future.

ImeevMA replied:
The problem is that these values are defined in different places. I think the best way would be to define them in the same place in the same enum, but not sure if that should be done in this patch-set. So far I have added two static asserts in two places. Also, I increased value of DT_IVAL_TO_STRING_BUFSIZE to 256, since 96 is too small.

Gerold103 reviewed (changes requested):
Thanks for the patch! I will continue the review later. Now some first comments. I looked through first 2 commits, they look good except these nits.

Gerold103 commented on extra/exports:
```diff
@@ -362,6 +362,7 @@ lua_upvalueid
 lua_upvaluejoin
 lua_xmove
 lua_yield
+interval_unpack
```
`i` goes before `l` (the file should be sorted alphabetically). But what is more important - we shouldn't add new symbols without `tnt_` prefix. Otherwise they might clash with user's symbols. The old code will have to live with the old names, but in all the new code we need `tnt_`.

ImeevMA replied:
Thanks, fixed.

Gerold103 commented on test/box-luatest/interval_test.lua:
```diff
@@ -0,0 +1,137 @@
+    g.server:exec(function()
+        local t = require('luatest')
+        local msgpack = require('msgpack')
+        local valbin = '\xc7\x07\x06\x03\x00\x01\x03\x7f\x05\x0f'
```
Can't you use `interval.new()`? Same in the other places below.

ImeevMA replied:
Fixed. However, we can't compare two intervals directly, so I still use `msgpack.encode()` to compare them.

ImeevMA replied:
Now we can use INTERVAL values in `assert_equals()`, so I dropped unnecessary code.

ImeevMA commented on src/lib/core/mp_interval.c:
```diff
@@ -0,0 +1,200 @@
+			value_size(itv->day) + value_size(itv->week) +
+			value_size(itv->month) + value_size(itv->year);
+	if (itv->adjust != DT_EXCESS)
+		size += 1 + mp_sizeof_uint((uint32_t)itv->adjust);
```
Thanks, fixed.

ImeevMA commented on src/lib/core/mp_interval.c:
```diff
@@ -0,0 +1,200 @@
+ * @param data A buffer.
+ * @param itv An interval to encode.
+ *
+ * @return data + mp_sizeof_interval(itv).
```
Dropped.

ImeevMA commented on src/lib/core/mp_interval.c:
```diff
@@ -0,0 +1,200 @@
+			return NULL;
+		}
+	}
+	*itv = tmp;
```
I used it to not set 0 for every field before I start decoding of MsgPack. Now, however, this doesn't make sense, as now some fields are of type DOUBLE, so it's better to set each field to 0 manually. Fixed.

ImeevMA commented on src/lib/core/mp_interval.c:
```diff
@@ -0,0 +1,200 @@
+		return -1;
+	char *str = tt_static_buf();
+	interval_to_string(&itv, str, TT_STATIC_BUF_LEN);
+	return snprintf(buf, size, "%s", str);
```
Thanks, fixed.

ImeevMA commented on src/box/lua/serialize_lua.c:
```diff
@@ -867,6 +867,12 @@ dump_node(struct lua_dumper *d, struct node *nd, int indent)
+			nd->mask |= NODE_QUOTE;
+			str = buf;
+			interval_to_string(field->interval, buf, sizeof(buf));
+			len = strlen(str);
```
Fixed.

ImeevMA commented on test/app-luatest/interval_test.lua:
```diff
@@ -0,0 +1,25 @@
+    t.assert_equals(res.min, 15)
+    t.assert_equals(res.sec, 0)
+    t.assert_equals(res.nsec, 0)
+    t.assert_equals(res.adjust, 1)
```
This is the default value for INTERVALs, which were obtained as the difference between two DATETIME values.

ImeevMA commented:
I've added another commit that introduces the `totable()` method for INTERVAL values. We can now use INTERVAL values in `require('luatest').assert_equals()`, so I've simplified the Luatest-related tests.

Gerold103 commented on third_party/lua-yaml/lyaml.cc:
```diff
@@ -618,7 +618,9 @@ static int dump_node(struct lua_yaml_dumper *dumper)
-   char buf[MAX(FPCONV_G_FMT_BUFSIZE, DT_TO_STRING_BUFSIZE)];
+   char buf[DT_IVAL_TO_STRING_BUFSIZE];
+   static_assert(DT_IVAL_TO_STRING_BUFSIZE > FPCONV_G_FMT_BUFSIZE);
+   static_assert(DT_IVAL_TO_STRING_BUFSIZE > DT_TO_STRING_BUFSIZE);
```
Static assert needs a message. Otherwise compilation works not everywhere.

ImeevMA replied:
Fixed.

Gerold103 reviewed (changes requested):
Good job on the fixes, only minor ones are left!

ImeevMA commented on src/lua/datetime.lua:
```diff
@@ -1114,12 +1114,32 @@ ffi.metatype(datetime_t, {
     __index = datetime_index,
```
I believe there was a doc-request about field of INTERVAL: in this [PR](https://github.com/tarantool/tarantool/pull/6923) we have commit [datetime: interval changelog and doc request](https://github.com/tsafin/tarantool/commit/669204071aa0a4219d87e136246b5d12b0722031).

Gerold103 replied:
But the doc request doesn't mention that the interval is an object which has these fields. The fields are described as arguments for sub/add/new. It means that if in the future we will decide to store hours mins and secs, for example, all as secs, how will `totable()` work then?
I don't mind keeping it but addition of a new public method just for internal testing I would rather recommend to avoid. Otherwise you will need to support it forever regardless of how the interval struct will mutate.

ImeevMA commented on src/lib/core/mp_interval.c:
```diff
@@ -0,0 +1,197 @@
+	itv->hour = 0.0;
+	itv->min = 0.0;
+	itv->sec = 0.0;
+	itv->nsec = 0;
```
Fixed.

ImeevMA commented on test/unit/interval.c:
```diff
@@ -0,0 +1,110 @@
+	mp_encode_interval(to_write, &itv);
+	const char *to_read = buf;
+	mp_decode_interval(&to_read, &result);
+	is(is_interval_equal(&itv, &result), true, "Intervals are equal.");
```
Thanks, fixed.

ImeevMA commented on test/unit/interval.c:
```diff
@@ -0,0 +1,110 @@
+	mp_encode_interval(to_write, &itv);
+	to_read = buf;
+	mp_decode_interval(&to_read, &result);
+	is(is_interval_equal(&itv, &result), true, "Intervals are equal.");
```
Thanks, applied.

Gerold103 reviewed (approved):
All looks good except for `:totable()` being unnecessary in the public namespace I think. Until somebody asks for it.

ImeevMA commented:
Thank you for the review! You are right, I dropped doc-request for `totable()`.
